pow/src/main/java/tech/pegasys/artemis/pow/DepositProcessingController.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.pow;
+
+import static tech.pegasys.artemis.pow.Eth1Manager.compareBlockTimestampToMinGenesisTime;
+import static tech.pegasys.artemis.pow.Eth1Manager.postMinGenesisTimeBlock;
+import static tech.pegasys.artemis.util.config.Constants.ETH1_FOLLOW_DISTANCE;
+
+import com.google.common.primitives.UnsignedLong;
+import io.reactivex.disposables.Disposable;
+import java.math.BigInteger;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.web3j.protocol.core.methods.response.EthBlock;
+import tech.pegasys.artemis.pow.api.Eth1EventsChannel;
+import tech.pegasys.artemis.util.async.AsyncRunner;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class DepositProcessingController {
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Eth1Provider eth1Provider;
+  private final Eth1EventsChannel eth1EventsChannel;
+  private final AsyncRunner asyncRunner;
+  private final DepositsFetcher depositsFetcher;
+
+  private volatile Disposable newBlockSubscription;
+  private boolean active = false;
+
+  // BlockByBlock mode is used to request deposit events and block information for each block
+  private boolean isBlockByBlockModeOn = false;
+  private volatile BigInteger blockByBlockLastFetchedBlockNumber = BigInteger.ZERO;
```

Comment:
We've got a weird mix of volatile and non-volatile fields in this class.  I think we'll wind up with calls from two sources - `Eth1Manager` and the latest block subscription.  Probably the best thing it to make all fields non-volatile but all entry point methods (including ones passed as handlers to the subscription `subscribe` method) should be `synchronised`.

Not 100% sure of that though.